### DISCUSSION TITLE
'Official' Docker for Beets

### DISCUSSION THREAD

**Discussion by bossanova808** [beetbox/beets#4864] [Ideas]
The LinuxServer.io beets docker is popular, but not well maintained.

As per this PR discussion: https://github.com/beetbox/beets/pull/4859#discussion_r1275564736 - I'd argue beets is exactly the sort of application that should have its own, internally maintained, docker.  Relying on a rather erratic third part organisation is not ideal, and we probably all know a beets installation is soon a fragile thing over time, with a lot of external dependencies (like replaygain tools) and issues with version (python libraries and whatnot).

Not only would an in house docker provide an easier path for most to a reliable, fully working install, but it would perhaps also help define 'best case' scenarios for tools and ways of using beets.  I think this would have significant benefits for beets as it takes only a casual glance through the github issues, discussions and the discourse to see issues in these area arising over and over again.

I am not per se volunteering to do it or maintain it, although I am also not ruling that out.  But I am far, far from an expert on either docker or beets...I just use them both reasonably regularly.  

The discussion should probably start with what the desirable outcomes would actually be.  The actual creating of the docker build and initial testing of it is the bulk of the work - once made, the maintenance requirements should be quite low (if other similar projects provide a guide, at least, e.g. https://hub.docker.com/r/lmscommunity/logitechmediaserver is one such project, in a not entirely unrelated area).

I'd think there are two basic approaches folks are likely to want/use:

1. **Basic Processing Container** 
(Like the existing LinuxServer beets - https://docs.linuxserver.io/images/docker-beets and https://www.linuxserver.io/blog/2016-10-08-managing-your-music-collection-with-beets) - i.e. a container that treats beets like a more typical media command line tool just for transient processing of new music - taking in a bunch of input and spitting out processed files from that (this is, for example, what I currently do with the linuxserver container).

2. **A container designed for ongoing, whole-library management use.** 
This would support more classic/intended use of beets as a full, ongoing library management tool.

The key things to work out, I guess, be:

- What's the right starting environment (lots of docker images seem to worry very much about smaller images, but I'd argue the best starting point would be a base that is very easy to maintain) - Alpine is popular for its smaller size, but Debian seems a good choice for maintenance and flexibility?  (I think LMS went with Debian for this reason).
- What are the best requirements for running beets in this environment - i.e. what, minimally, on top of base Debian is needed
   e.g. Python version, needed external libraries for core functionality etc.
- What are the key/most popular plugins and what do they need?  These, I think, need to be included by default.
- Should _all_ plugins be included, or should there perhaps be an easy way of adding further plugins?

Personally, other than the need to maintain it, I can't see how having an official containerised version of beets would not be of great utility in helping folks discover and use beets more successfully.

What do other folk think?

**Comment by sampsyo** [comment]
I fully encourage some effort on this!

The main recommendation I'd have is to start small and iteratively improve. The question above about whether to include dependencies for "all plugins" is an important one; I strongly think we should start with a minimal Dockerfile that includes dependencies for _no_ plugins and just runs the "base" beets. The nice thing about Docker containers is that we can build a more complete container that builds on the base one as a layer. So by starting minimally, we can have something useful for a few users while setting us up to serve more users in the future.

**Comment by RollingStar** [comment]
You probably already saw, but this is my experience with the linuxserver beets container (scroll down).

https://discourse.beets.io/t/diary-of-beets-on-arm-odroid-hc4-armbian/1993

Possibly related to running on arm64. I had to compile some stuff but otherwise haven't had an issue so far.

I'm pretty technical, and I _use_ docker but don't really _understand_ it. Specifically, I don't understand how to customize images/dockerfiles.

https://linuxconfig.org/how-to-customize-docker-images-with-dockerfiles

This is a better writeup than most of what I've seen.

---

If someone was going down this path, I'd recommend talking to experts and determining the support footprint of docker vs. other means as a maintainer and an end user. Docker on paper has a lot of benefits as a user. As a maintainer can we ever get to only supporting docker? what's the infra like on windows? quick google implies options are 'docker desktop' , WSL, and maybe a third docker option? https://dev.to/_nicolas_louis_/how-to-run-docker-on-windows-without-docker-desktop-hik

doesn't give me hope.

Are the end user benefits real or theoretical? A lot of my containers seem to be outdated and black boxy. I've been using portainer and it looks like updating containers in the free version is up to the end user 😳

What perf penalty does docker suffer? I believe musicbrainz uses docker for the server (at least the run-at-home server) so hopefully random db read/writes aren't a big strain.

**Comment by bossanova808** [comment]
Most of those are relatively simple questions...Docker is massively well established and supported at this point, and works very well across many platforms (kind of the whole point), and docker performance is very high, there's very little impact in most scenarios. 

E.g. even Docker on Windows is very well established and works very well.  I use docker in WSL2 on a daily basis for web dev work (using a great project called DDEV - https://ddev.com/) , and it's excellent and very performant.   This is using docker simply installed inside of WSL2, but I have also use Docker Desktop in the past before this was an option.  In both cases, then running beets inside of this docker should work just fine, given other very complex apps/containers/systems of containers do.   There is one proviso - there are some disk speed issues with crossing the Linux <> windows disk boundaries, if this is needed.  But there are a variety of workable ways around that, in this context, depending on what one is trying to do.

Further, in many web dev cases, people then carry on to use docker containers for services on their actual live server (including the web server, PHP or other server language, DB etc., all as separate containers) - as the performance impact is so small, and the benefits of containerisation so great.

So I think one can dismiss pretty much all basic questions of platform and performance (with the proviso that if the hardware you're on is somehow fundamentally hamstrung (low power CPU, slow bus connection to the data volumes etc) - that will affect docker just as much as a local install, and indeed probably marginally more.  Of course the beets image would want to be multi-architecture to allow running on a good variety of systems.

I'm not sure you'd ever want to 'only' support docker, though.   I suspect you might want to support it as the _default_ way of running beets, but of course it still has to run in various other ways.  In this case it's really just about controlling versions & dependencies and providing a 'best case' install.

One of the main points of containerisation is actually to make things more 'black boxy' - and along with that to thus provide stability and more control over updates/upgrades.  It's certainly usually considered bad practise, generally, in the docker world, to automate container updates.  There are various management things (like Portainer, although I personally use DIUN (I think the LinuxServer folks recommend DIUN actually https://crazymax.dev/diun/ - in itself a dockerised service) - which will alert one to the _availability_ of an update, but the actual update should always be a deliberate user decision with due care.  In general, with services, you don't usually want or need the very latest, bleeding edge, stuff - but with a docker approach it's easy to e.g. run a service in a known stable form whilst experimenting with second, more up to date container.

**Comment by Shredder5262** [comment]
I have just installed the linuxserver image and and nearing the end of importing my music library and my experience has been....it's slow AF, in comparison to running it in a python/windows environment where returned output was almost instant.. Everything takes at least 5-10 minutes to do after entering a command. I have been wondering if there are silent errors going on, but this is just really sluggish...even using the info command is slow. So far I'm not impressed by putting this in a docker container.................that said If there is a better image out there that improves on performance... you've got my vote on trying it. 

What I like about having it in docker is when it comes to sideloading things like imagemagick and other plugins like filetote is significantly easier to deal with...if you know what you're doing. Just plop the commands in the docker compose file and all is well. Not having to deal with windows environment variables is great.  It will be easier to automate beets container updates ( I know it's bad practice....I don't care) and also library updates.

**Reply by snejus** [reply]
Does the slowness affect the other images as well, or just the beets image?

**Reply by Shredder5262** [reply]
I notice a sluggishness in general with using docker and the darr stuff, but it's not anything intolerable, beets is the only one that I have noticed a significant slowness. 

At the time of writing this, my host machine is...
 7-13% CPU
46-56% Ram Usage
68% SWAP Ram Usage
So, my host machine isn't really struggling, 
Docker resources are 
14% cpu
13GB of 31GB of  Ram
21 of 29 containers runnning....so no issue there imo.... don't know why it's slow....maybe network...this is communicating with a nas nearly everything is running through a single bridge connection

If i had to throw a dart at a dartboard and guess.... I would say that it's silently erroring somewhere....but I don't know where. I've tried running with verbose output for a while...nothing glaringly obvious....my config may or may not be extensive (In my opinion it's not anything outside the wildness i've seen with other configs...so) 

I've just recently tried this filetote plugin that, in my experience, is very hit and miss...perhaps i don't have it set up right, but that might be something......I don't know...it's just really slow for me. 

Here is my config ...
`plugins: filetote fetchart embedart convert scrub replaygain lastgenre chroma web info ftintitle inline missing zero edit unimported discogs types bucket mbsync
directory: /music
library: /config/musiclibrary.blb
art_filename: thumb
threaded: yes
original_date: no
per_disc_numbering: yes
sort_album: albumartist+ album+
sort_item: artist+ album+ disc+ track+
sort_case_insensitive: no  
asciify_paths: no
replace:
  '^\.': _
  '[\x00-\x1f]': _
  '[<>:"\?\*\|]': _
  '[\xE8-\xEB]': e
  '[\xEC-\xEF]': i
  '[\xE2-\xE6]': a
  '[\xF2-\xF6]': o
  '[\xF8]': o
  '\.$': _
  '\s+$': ''
  '[\\/]': _
  
convert:
  auto: no
  ffmpeg: /usr/bin/ffmpeg
  opts: -ab 320k -ac 2 -ar 48000
  max_bitrate: 320
  threads: 1
    
item_fields:
    multidisc: 1 if disctotal > 1 else 0

paths:
    kind:Game: Soundtrack/VideoGame/$year - $album %aunique{}/$albumartist - $album ($year) - %if{$multidisc,$disc-} Track $track - $title - $format - $bitrate
    kind:Movie: Soundtrack/Movie/$year - $album %aunique{}/$albumartist - $album ($year) - %if{$multidisc,$disc-} Track $track - $title - $format - $bitrate
    kind:Anime: Soundtrack/Anime/$year - $album %aunique{}/$albumartist - $album ($year) - %if{$multidisc,$disc-} Track $track - $title - $format - $bitrate
    albumtype:soundtrack: Soundtrack/$year - $album %aunique{}/$albumartist - $album ($year) - %if{$multidisc,$disc-} Track $track - $title - $format - $bitrate
    comp: Compilations/$year - $album %aunique{}/$albumartist - $album ($year) - %if{$multidisc,$disc-} Track $track - $title - $format - $bitrate
    singleton: Non-Album/$albumartist/$year - $album %aunique{}/$albumartist - $album ($year) - %if{$multidisc,$disc-} Track $track - $title - $format - $bitrate
    default: Artists/%bucket{$albumartist,alpha}/$albumartist/$year - $album %aunique{}/$albumartist - $album ($year) - %if{$multidisc,Disc $disc -}Track $track - $title - $format - $bitrate
    
    
aunique:
    keys: albumartist album
    disambiguators: album albumtype year label catalognum albumdisambig releasegroupdisambig
    bracket: '[]'
    
import:
  write: yes
  move: yes
  resume: ask
  incremental: yes
  quiet_fallback: skip
  timid: no
  log: /config/beet.log

filetote:
  auto: yes
  log: true
  extensions: .png .jpg .jpeg
  ignore_case: yes
  rmdir: yes       # Remove empty directories after moving
  patterns:
    artworkdir:
      - "(?i)artwork/"
    images:
      - "(?i).*back.*"
      - "(?i).*spine.*"
      - "(?i).*discart.*"
      - "(?i).*thumb.*"
      - "(?i).*cover.*"
      - "(?i).*folder.*"
      - "(?i).*front.*"
      - "(?i).*inlet.*"
      - "(?i).*cd.*"
      - "(?i).*disc.*"
embedart:
  auto: yes
  remove_art: no

fetchart:
  auto: yes
  cautious: true
  cover_names: thumb cover front discart art album folder disc spine back Booklet Other Medium 
  minwidth: 500
  enforce_ratio: yes
  fanarttv_key: [redacted]
  lastfm_key: [redacted]
  google_key: [redacted]
  google_engine_ID: 009217259823014548361:lndtuqkycfu
  sources: 
    - filesystem
    - coverart: release
    - fanarttv 
    - lastfm
    - coverart
    - itunes
    - coverart: releasegroup
    - amazon
    - musicbrainz
    - albumart
    - google
    
# Auto Tagger ----------------------------------------------------------------    
match:
  strong_rec_thresh: 0.04
  medium_rec_thresh: 0.25
  rec_gap_thresh: 0.25
  max_rec:
     missing_tracks: medium
     unmatched_tracks: medium
  ignored_media: ['Data CD', 'DVD', 'DVD-Video', 'Blu-ray', 'HD-DVD',
                 'VCD', 'SVCD', 'UMD', 'VHS', 'video']

lyrics:
  auto: yes
  dist_thresh: 0.11
  fallback: null
  force: no
  sources: [lrclib, google, genius]
  synced: yes 
    
lastgenre:
  auto: yes
  count: 1
  force: yes
  keep_existing: no
  source: track
  whitelist: /config/genrewhitelist.txt
  
ftintitle:
  auto: yes
  format: (feat. {0})
  
replaygain:
  auto: no

scrub:
  auto: yes
  
missing:
  format: $albumartist - $album - $track - $title
  count: no
  total: no

zero:
  fields: comments

unimported:
  ignore_extensions: jpg png lrc txt nfo JPG

discogs:
  user_token: [redacted]
  index_tracks: yes


      
bucket:
  bucket_alpha: ['0-9', 'A-C', 'D-F', 'G-I', 'J-L', 'M-O', 'P-R', 'S-U', 'V-Z']
  bucket_alpha_regex:
    '0-9': '^[^A-Za-z0-9]*[0-9]'       # Ignore leading non-alphanumeric chars, then match digit
    'A-C': '^[^A-Za-z0-9]*[A-Ca-c]'    # Ignore leading special chars, match A-C
    'D-F': '^[^A-Za-z0-9]*[D-Fd-f]'
    'G-I': '^[^A-Za-z0-9]*[G-Ig-i]'
    'J-L': '^[^A-Za-z0-9]*[J-Lj-l]'
    'M-O': '^[^A-Za-z0-9]*[M-Om-o]'
    'P-R': '^[^A-Za-z0-9]*[P-Rp-r]'
    'S-U': '^[^A-Za-z0-9]*[S-Us-u]'
    'V-Z': '^[^A-Za-z0-9]*[V-Zv-z]'

web:
  host: 0.0.0.0
  port: 8337


clutter:
  - ".DS_Store"
  - "Thumbs.db"
  - "*.nfo"
  - "*.txt"
  - "*.log"
  - "*.@__thumb*"
  - "(?i).*\\.nfo$"
  - "(?i).*\\.log$"
  - "(?i).*\\.cue$"
  - "(?i)^\\.DS_Store$"
  - "(?i).*\\.jpg$"
  - "*.m3u"
  - "*.html"
  - "*.htm"
  - "*.nfo"
  - "*.log"
  - "*.url"
  - "*.cue"
  - "*.nzb"
  - "*.php"
  - "*.biz"
  - "*.txt"
  - "*.srr"`

Docker compose file
`---
services:
  beets:
    image: lscr.io/linuxserver/beets:latest
    container_name: beets
    environment:
      - PUID=1000
      - PGID=1000
      - TZ=America/Chicago
      - DOCKER_MODS=linuxserver/mods:universal-package-install
      - INSTALL_PACKAGES=imagemagick libjpeg-turbo libpng imagemagick-jpeg
      - INSTALL_PIP_PACKAGES=beets-filetote
    volumes:
      - network_multimedia:/music #optional
      - host machine/Beets-data:/config
      - host path\Music - Lidarr\:/downloads #optional
      - host path\Music - Lidarr\:/musicrecyclebin
    ports:
      - 8337:8337
    restart: unless-stopped
    networks:
      - darr_network
volumes:
  network_multimedia:
    driver_opts:
      type: nfs
      o: addr=nas address,rw,vers=3
      device: ":nas music path"
networks:
 darr_network:
    external: true`

That's about everything with the way i set it up...

**Comment by bossanova808** [comment]
Something wrong there.  I use the linuxserver docker all the time and it's plenty quick, quite comparable to direct use...

**Reply by Shredder5262** [reply]
Curious, how big is your db?

**Comment by Shredder5262** [comment]
I thought I would pass this along as valuable info, running docker desktop on a windows host and previous posts complaining about slowness.... I found that if you mount the beets db as a linux volume then it provides a SIGNIFICANT speed difference

Below is my docker compose file (redacted), but the important lines are 
      - beets_db:/db                         # Docker volume just for the DB
and
  beets_db: {}   # <-- Docker-managed volume for fast SQLite DB

This creates 2 seperate folders in windows, a mounted volume for linux...and a more windows friendly path... I store my config in the windows friendly path for quick config editing. 

Also, in the config, I have 

pluginpath: /config/beetsplug
library: /db/musiclibrary.blb

To remap the music library to the mounted volume. 

If using docker on a windows system...this is the way. 


```
services:
  beets:
    image: 
    container_name: beets
    environment:
      - PUID=1000
      - PGID=1000
      - TZ=America/Chicago
      - DOCKER_MODS=linuxserver/mods:universal-package-install
      - INSTALL_PACKAGES=imagemagick libjpeg-turbo libpng imagemagick-jpeg
      - INSTALL_PIP_PACKAGES=beets-filetote
    volumes:
      - network_multimedia:/music
      - configpath :/config   # Windows path for config.yaml
      - beets_db:/db                         # Docker volume just for the DB
      - music download folder\:/downloads
      - music recycle bin\:/musicrecyclebin
    ports:
      - 8337:8337
    restart: unless-stopped
    networks:
      - docker_network_name

volumes:
  network_multimedia:
    driver_opts:
      type: nfs
      o: addr=my nasdrive \rw,vers=3
      device: ":/musicpath/"
  beets_db: {}   # <-- Docker-managed volume for fast SQLite DB

networks:
  docker_network_name:
    external: true

```

**Comment by bossanova808** [comment]
It's very well known that with WSL2 the performance is night and day between files in the Linux filesystem vs. the Windows host.  Of course for the db this would be critical, but frankly the music access would be a big thing too I am sure.  You'd be far better off having it all over on the Linux side of things, if speed is a concern.

Frankly, unless it's for small/basic things, WSL2 accessing the Windows side of things is just untenably slow, across the board.

Oh and sorry I missed your question above.  My DB is around 800 artists, with over 2000 albums.  I run this all on Linux (well, Proxmox, with a Debian 13 LXC, in turn running Docker, with the Linuxserver beets...).  No speed issues at all (well apart from with Musicbrainz queries being sluggish, but that's a general Musicbrainz issue that one experiences with Picard etc as well).
